nemo_aligner/algorithms/ppo.py

```diff
@@ -201,84 +270,121 @@ def post_process_tensor(tensor):
+
+            unbalanced_rm_value_batch = PPORolloutBatch.from_rollout_batches(
+                rm_value_rollout_batches,
+                eos_id=self.model.tokenizer.eos_id,
+                rollout_batch_seq_length=padded_rollout_sequence_length,
+            )
+            global_rm_value_batch = unbalanced_rm_value_batch.gather_and_balance_globally()
 
-        tensor_to_accumulate = torch.tensor(
-            [metrics["response_lengths"], metrics["prompt_lengths"], metrics["rewards"], num_samples],
-            dtype=torch.float32,
-            device=torch.cuda.current_device(),
+        # chunking needs to be outside of reshard region
+        # NOTE: the seed here must be the same as the chunk before since we need to shuffle
+        # these values the same way as the other values
+        balanced_rm_value_batch = global_rm_value_batch.chunk(
+            parallel_state.get_data_parallel_rank(), parallel_state.get_data_parallel_world_size(), self.step
         )
-        torch.distributed.all_reduce(tensor_to_accumulate, group=parallel_state.get_data_parallel_group())
+        balanced_local_batch.update(balanced_rm_value_batch)
+
+        global_rollout_batch.update(global_rm_value_batch)
+
+        return balanced_local_batch, cpu_dict(self.compute_rollout_metrics(global_rollout_batch)), timer_metrics
+
+    def compute_rollout_metrics(self, rollout_batch):
+        table = {}
 
-        (
-            global_response_lengths,
-            global_prompt_lengths,
-            global_rewards,
-            global_num_samples,
-        ) = tensor_to_accumulate.tolist()
+        prompt_lengths = rollout_batch["prompt_lengths"]
+        response_lengths = rollout_batch["response_lengths"]
+        response_tokens = rollout_batch["response_tokens"]
+        rewards = rollout_batch["rewards"]
+        is_end = rollout_batch["is_end"]
+
+        reward = rewards[0]
+        prompt_length = prompt_lengths[0]
+        response_length = response_lengths[0]
+        response_token = response_tokens[0]
+
+        table["reward"] = reward.item()
+        table["prompt"] = self.model.tokenizer.ids_to_text(response_token[:prompt_length].tolist())
+        table["response"] = self.model.tokenizer.ids_to_text(response_token[prompt_length:response_length].tolist())
 
         metrics = {
             "table": table,
-            "global_response_lengths_mean": global_response_lengths / global_num_samples,
-            "global_prompt_lengths": global_prompt_lengths / global_num_samples,
-            "global_rewards": global_rewards / global_num_samples,
+            "consumed_samples": prompt_lengths.size(0),
+            "response_lengths": response_lengths.float().mean().item(),
+            "prompt_lengths": prompt_lengths.float().mean().item(),
+            "generation_length": (response_lengths - prompt_lengths).float().mean().item(),
+            "rewards": rewards.mean().item(),
+            "amount_of_samples_properly_ended": is_end.sum().item(),
```

Comment:
Can we make it a % instead? Something like
```suggestion
            "fraction_of_samples_properly_ended": is_end.float().mean().item(),
```